Add System.Composition

dsplaisted opened the pull request:
This PR adds System.Composition to CoreFx.

System.Composition consists of 5 different assemblies:

- System.Composition.AttributedModel
- System.Composition.Convention
- System.Composition.Hosting
- System.Composition.Runtime
- System.Composition.TypedParts

There is a general test project for all the assemblies, and another one specific to System.Composition.Convention.  There are also two assemblies referenced by the general test project, and a project for perf testing.

I've created separate folders under /src for each of the five main assemblies, and put the general test project, the assemblies it references, and the perf test project in /src/System.Composition.  I've put the solution file (System.Composition.sln) directly under /src.

I think this follows the current repository structure as closely as possible.  Perhaps it makes more sense to put everything here under a System.Composition folder.  Feedback is welcome.

## Remaining issues
- [x] Are AssemblyInfo.cs files needed?
- [x] Remove unused usings?
- [x] Dependencies in project.json should be consistent with rest of repo (ie use "-beta-*")

## Follow-up for after PR is merged
- [ ] Clean up FEATURE_TRACING.
- [ ] Get rid of `ReadLock` type ([discussion](https://github.com/dotnet/corefx/pull/1851#discussion_r31379745))
- [ ] Address commented out tests in ConventionBuilderTests.cs
- [ ] Add documentation around CompositionOperation, lifetime, and threading ([discussion](https://github.com/dotnet/corefx/pull/1851#discussion_r31377139))
- [ ] Should strings in LifeTimeContext.ToString be localized?
- [ ] Use Debug.Assert instead of Requires? ([discussion](https://github.com/dotnet/corefx/pull/1851#discussion_r31377139))
- [ ] Consider removing ExceptionBuilder ([discussion](https://github.com/dotnet/corefx/pull/1851#discussion_r31249633))
- [ ] Uncomment or delete commented out localizable attributes ([discussion](https://github.com/dotnet/corefx/pull/1851#discussion_r36592804))

stephentoub commented on src/System.Composition.AttributedModel/src/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,18 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Resources;
```
Are such AssemblyInfo.cs files needed?  I was under the impression that the buildtools supplied such attributes automatically.

stephentoub commented on src/System.Composition.AttributedModel/src/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,18 @@
+[assembly: NeutralResourcesLanguage("en")]
+
+[assembly: ComVisible(false)]
+[assembly: SecurityTransparent]
```
We had problems in other libraries when [SecurityTransparent] was applied, as the rest of the libraries in corefx are the default of non-transparent, and so any attempts to use those from this (e.g. Debug.Assert) will fail.

dsplaisted replied:
@stephentoub Transparency checks [have been disabled](https://github.com/dotnet/coreclr/commit/725d6a91fffdf4603c8a9c74e15cbc83ebf0e3d6) in .NET Core.  Adding System.Composition to .NET Core was actually blocked on this, as there was an issue where dynamically compiled code would run into security exceptions no matter what transparency attributes were applied.

However, I would suggest leaving the transparency attributes in this code base as is, to reduce the chance of breaking something on the other platforms this library supports.  I think Windows Phone 8/8.1 would by risky here.

stephentoub commented on src/System.Composition.AttributedModel/src/System.Composition.AttributedModel.csproj:
```diff
@@ -0,0 +1,66 @@
+    <SccProjectName>SAK</SccProjectName>
+    <SccLocalPath>SAK</SccLocalPath>
+    <SccAuxPath>SAK</SccAuxPath>
+    <SccProvider>SAK</SccProvider>
```
Seems like a lot of this stuff can be removed; might be good to look at .csproj files elsewhere in the repo as a guide.

dsplaisted replied:
@stephentoub Thanks, I've cleaned up the project files.

stephentoub commented on src/System.Composition.AttributedModel/src/System.Composition.AttributedModel.csproj:
```diff
@@ -0,0 +1,66 @@
+  <Target Name="AfterBuild">
+  </Target>
+  -->
+</Project>
```
Nit: some files are missing newlines at the end.

dsplaisted replied:
@stephentoub  Does this matter at all?  Does this apply just to project files or also to source files?  Should this be part of the code formatter?

stephentoub replied:
> Does this matter at all? 

Yes, some example discussions of the topic:
- http://stackoverflow.com/questions/729692/why-should-files-end-with-a-newline
- http://stackoverflow.com/questions/5813311/no-newline-at-end-of-file
- http://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file
- https://robots.thoughtbot.com/no-newline-at-end-of-file

> Does this apply just to project files or also to source files? 

All text-based files.

> Should this be part of the code formatter?

That would help, but codeformatter doesn't process all files for which this is relevant.

dsplaisted replied:
OK, I've added newlines to all the .csproj and .cs files.  I've also [submitted a PR](https://github.com/dotnet/codeformatter/pull/127) to the code formatter tool to automatically add newlines to the end of C# files.

stephentoub commented on src/System.Composition.AttributedModel/src/System/Composition/Convention/AttributedModelProvider.cs:
```diff
@@ -0,0 +1,33 @@
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
```
Looks like there are some unnecessary usings, e.g. System.Threading.Tasks.

dsplaisted replied:
@stephentoub It doesn't look like removing unnecessary using statements is explicitly part of the [Coding style guidelines](https://github.com/dotnet/corefx/wiki/Coding-style).  Is this something we want to add?

stephentoub replied:
I view unused using as dead code, and dead code should be removed.  Doesn't seem like a style thing to me, personally.

stephentoub commented on src/System.Composition.AttributedModel/src/System/Composition/ExportMetadataAttribute.cs:
```diff
@@ -0,0 +1,59 @@
+        public string Name
+        {
+            get;
+            private set;
```
Looks like some inconsistencies in formatting between files; the previous file had all of the property's definition on one line.

dsplaisted replied:
@stephentoub Most of the auto-implemented properties were on one line, so I've formatted the ones that weren't that way.

stephentoub commented on src/System.Composition.AttributedModel/src/System/Composition/PartMetadataAttribute.cs:
```diff
@@ -0,0 +1,58 @@
+        public string Name
+        {
+            get;
+            private set;
```
In some of these attributes, private readonly fields are being used as the backing store, whereas in others private setters are being used on the properties.  Is there a rhyme/reason to when each approach is used?

stephentoub commented on src/System.Composition.Convention/src/Microsoft/Composition/Diagnostics/CompositionTraceSource.cs:
```diff
@@ -0,0 +1,58 @@
+{
+    internal static class CompositionTraceSource
+    {
+#if FEATURE_TRACING
```
Are both cases of FEATURE_TRACING and !FEATURE_TRACING still needed?

stephentoub commented on src/System.Composition.Convention/src/Microsoft/Composition/Diagnostics/SilverlightTraceWriter.cs:
```diff
@@ -0,0 +1,88 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
Should the file be renamed to DebuggerTraceWriter.cs?

stephentoub commented on src/System.Composition.Convention/src/Microsoft/Composition/Diagnostics/SilverlightTraceWriter.cs:
```diff
@@ -0,0 +1,88 @@
+
+        public override void WriteWarning(CompositionTraceId traceId, string format, params object[] arguments)
+        {
+            WriteEvent(TraceEventType.Warning, traceId, format, arguments);
```
If CanWriteWarning and CanWriteError are hard coded to return false, should these WriteWarning and WriteError methods instead throw a NotSupportedException?

stephentoub commented on src/System.Composition.Convention/src/Microsoft/Composition/Diagnostics/SilverlightTraceWriter.cs:
```diff
@@ -0,0 +1,88 @@
+
+        private static void WriteEvent(TraceEventType eventType, CompositionTraceId traceId, string format, params object[] arguments)
+        {
+            if (true /* !Debugger.IsLogging() */)
```
Should this body just be empty?

stephentoub commented on src/System.Composition.Convention/src/Microsoft/Internal/Assumes.InternalErrorException.cs:
```diff
@@ -0,0 +1,23 @@
+    {
+        // The exception that is thrown when an internal assumption failed.
+        [SuppressMessage("Microsoft.Design", "CA1064:ExceptionsShouldBePublic")]
+        private class InternalErrorException : Exception
```
sealed?

stephentoub commented on src/System.Composition.Convention/src/Microsoft/Internal/Lock.Reader.cs:
```diff
@@ -0,0 +1,32 @@
+
+        public void Dispose()
+        {
+            if (Interlocked.CompareExchange(ref _isDisposed, 1, 0) == 0)
```
This is only effective if the struct isn't copied, as each independent copy will have its own _isDisposed.

KrzysztofCwalina replied:
Yeah, this seems like a very dangerous type.

nblumhardt replied:
Definitely could be a footgun in waiting. So long as usage is always as per:

```csharp
using (new ReadLock(_lock))
{
    _memberInfos.TryGetValue(memberInfo, out cachedAttributes);
}
```

...then it's benign, but switching this and `WriteLock` to `class`es would be a reasonable change as well.

This code when through some pretty heavy profiling/optimisation - the original goal was to get composition such as `new A(new B())` or `new A(B.Instance)` in line with the native/hand-coded equivalents. (Other IoC containers under development at the time were also heading for this goal.)

Garbage on the heap was a big barrier to to it - but without the harness/original numbers I can't say just how much. @dsplaisted are there any perf tests with this that might point to how much the switch to `class`es would cost?

dsplaisted replied:
@nblumhardt I added the perf tests that were part of the Roxel codebase under src/System.Composition/perftests.  Do you know if there were any other perf tests elsewhere?

nblumhardt replied:
Cool, I'd guess those are the only ones.

Thinking a bit more about how this is used (during reflection, which is done once only for each component type IIRC), I'd say the risk of any major regression making the change is pretty low.

stephentoub replied:
FWIW, I wasn't actually suggesting adding more cost by using a class.  Seems like you don't need the type at all, as:
```
_lock.EnterReadLock();
try { 
    _memberInfos.TryGetValue(memberInfo, out cachedAttributes);
}
finally { _lock.ExitReadLock(); }
```
seems just as clear as:
```
using(new ReadLock(_lock))
{
    _memberInfos.TryGetValue(memberInfo, out cachedAttributes);
}
```
and without any additional overhead.  Especially since there are only a handful of places internally where these are used.  If you really wanted to keep the struct, because you believe it does add to the simplicity and there are no concerns about misuse, then the Interlocked itself seems like unnecessary overhead and could just be removed: if we're relying on the consumer only using the right pattern, then trying to protect against one particular misuse doesn't seem warranted.

nblumhardt replied:
+1, had a similar thought a couple of hours after that last post :-)

@dsplaisted not sure how the process works; if this is merged and an issue raised for this item I'd love to help out with it. Ping me if/when!

stephentoub commented on src/System.Composition.Convention/src/Microsoft/Internal/Lock.cs:
```diff
@@ -0,0 +1,79 @@
+{
+    internal sealed class Lock : IDisposable
+    {
+#if FEATURE_SLIMLOCK
```
ReaderWriterLockSlim does exist in CoreCLR, so will FEATURE_SLIMLOCK always be defined, and if so, can we get rid of the rest of the dead code?

dsplaisted replied:
@stephentoub I've removed the dead code.

stephentoub commented on src/System.Composition.Convention/src/Microsoft/Internal/Lock.cs:
```diff
@@ -0,0 +1,79 @@
+    internal sealed class Lock : IDisposable
+    {
+#if FEATURE_SLIMLOCK
+        private ReaderWriterLockSlim _thisLock = new ReaderWriterLockSlim(LockRecursionPolicy.NoRecursion);
```
readonly?

stephentoub commented on src/System.Composition.Convention/src/Microsoft/Internal/Lock.cs:
```diff
@@ -0,0 +1,79 @@
+
+        public void Dispose()
+        {
+            if (Interlocked.CompareExchange(ref this._isDisposed, 1, 0) == 0)
```
Why protect against multiple Dispose calls?  Is the concern that code might be using Dispose concurrently?  If so, that seems like poorly written consuming code.  Otherwise, shouldn't Dispose allow for multiple calls to Dispose, where subsequent calls are nops?

stephentoub commented on src/System.Composition.Convention/src/Microsoft/Internal/Requires.cs:
```diff
@@ -0,0 +1,104 @@
+            }
+            Contract.EndContractBlock();
+        }
+        //    [DebuggerStepThrough]
```
Can we delete all of this commented-out code?

stephentoub commented on src/System.Composition.Convention/tests/ConventionBuilderTests.cs:
```diff
@@ -0,0 +1,202 @@
+
+        [Export(typeof(IGenericInterface<>))]
+        private class ClassExportingInterface<T> : IGenericInterface<T> { }
+        //[TestMethod]
```
Should this commented out code be removed, or do we have plans to uncomment them (in which case an issue should be opened)?

stephentoub commented on src/System.Composition.Hosting/src/Microsoft/Internal/Assumes.InternalErrorException.cs:
```diff
@@ -0,0 +1,23 @@
+    {
+        // The exception that is thrown when an internal assumption failed.
+        [SuppressMessage("Microsoft.Design", "CA1064:ExceptionsShouldBePublic")]
+        private class InternalErrorException : Exception
```
This and the related Assumes/Requires/etc. code is being duplicated across projects.  Can we not define them once and share those files, as we do with other Common code?

dsplaisted replied:
@stephentoub I've moved this code into the Common folder.

stephentoub commented on src/System.Composition.Hosting/src/System/Composition/Hosting/Core/CompositionOperation.cs:
```diff
@@ -0,0 +1,121 @@
+            if (_sharingLock == null)
+            {
+                _sharingLock = sharingLock;
+                Monitor.Enter(sharingLock);
```
I don't know exactly how this function is used, but this code seems very odd.  For example, if multiple threads race to call EnterSharingLock, then each thread could end up seeing _sharingLock as nul and calling Monitor.Enter on their own sharingLock object.

nblumhardt replied:
`CompositionOperation` instances are strictly single-threaded. (The class comment notes it with `<remarks>Instances of this class are not safe for access by multiple threads.</remarks>`.)

stephentoub replied:
Then why does it have such a locking mechanism?

nblumhardt replied:
Hey Stephen - here's the essential background, let me know if I can add more info. Will happily PR some documentation/comments that might be a bit lacking in places :-)

The core model for composition in this lib is a tree of `LifetimeContext` scopes, with longer lived scopes towards the root of the tree and short-lived ones towards the leaves. E.g. when representing components in a web application, the ones created in the root scope might be "singleton" and stick around for the life of the app, each concurrent web request might have a scope at the next level down for holding things like ORM sessions, and finally there might be some further nested scopes for managing the lifetimes of components used just for short transactions.

Composition operations build a graph of objects associated with a single `LifetimeContext` (usually starting at a leaf one). Sometimes these trigger more composition operations at other scopes up towards the root when instances of longer-lived components are required.

The scopes are shared at each level by the composition operations running on them, which is often done without contention, but when the operation hits a part whose instances are "shared" in that particular scope (e.g. all instances associated with a web request share the same ORM session) the sharing lock for that level is taken so that only one unique instance is created.

Rather than immediately releasing the lock, it has to be held for the duration of the composition operation's work in that scope, because some actions (e.g. property initialization) are deferred until other parts of the graph are created. The lock prevents shared parts from becoming visible to other composition operations until they're definitely fully constructed.

Hope this covers your question. All the best!

stephentoub replied:
Thanks for the additional details.  I'm still not following 100%.  To be clear, the same CompositionOperation instance will never be used concurrently, e.g. the same instance will never be passed to GetOrCreate at the same time on two different threads?  If so, why is the Monitor.Enter necessary at all?  The way this is constructed, if EnterSharingLock has already been called, then _sharingLock will forever be non-null, so any subsequent callers won't even invoke Monitor.Enter, so what's the purpose of the Enter call?

nblumhardt replied:
Apologies in advance if I head off down the garden path here, I could be misunderstanding your query.

Though the same instance will never be used concurrently, multiple `CompositionOperation` instances running on _different_ threads may be active on the same scope concurrently. The scope owns the `_sharingLock`.

The first `CompositionOperation` to take `_sharingLock` will hold it until that operation ends, after which (in `Dispose()`) it will relinquish the lock and another `CompositionOperation` will proceed if any are waiting.

What might be the crucial missing info:

During a composition operation, `EnterSharingLock()` is called before beginning construction of a shared part within  a scope, _irrespective of whether the lock is already held_ by that operation, hence the subsequent calls to `EnterSharingLock()` being no-ops at that point. The name of the method could change to better reflect this - e.g. `EnsureSharingLockIsHeld()` would be an improvement I think.

stephentoub replied:
Ah!  Ok, I get it now.  Thanks for the extra details.  And, yeah, comments explaining all this would be very welcome ;)

stephentoub commented on src/System.Composition.Hosting/src/System/Composition/Hosting/Core/LifetimeContext.cs:
```diff
@@ -0,0 +1,222 @@
+        public override string ToString()
+        {
+            if (_parent == null)
+                return "Root Lifetime Context";
```
These strings don't need to be localized?

stephentoub commented:
Lots of tests are failing, with errors like:
```
MESSAGE:
System.IO.FileNotFoundException : Could not load file or assembly 'System.Reflection.Emit.Lightweight, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
+++++++++++++++++++
STACK TRACE:
at System.Linq.Expressions.Compiler.LambdaCompiler..ctor(AnalyzedTree tree, LambdaExpression lambda) at System.Linq.Expressions.Compiler.LambdaCompiler.Compile(LambdaExpression lambda) at System.Linq.Expressions.Expression`1.Compile() at System.Composition.Hosting.Providers.Metadata.MetadataViewProvider.GetMetadataViewProvider[TMetadata]() in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Hosting\src\System\Composition\Hosting\Providers\Metadata\MetadataViewProvider.cs:line 73 at System.Composition.Hosting.Providers.Lazy.LazyWithMetadataExportDescriptorProvider.GetLazyDefinitions[TValue,TMetadata](CompositionContract lazyContract, DependencyAccessor definitionAccessor) in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Hosting\src\System\Composition\Hosting\Providers\Lazy\LazyWithMetadataExportDescriptorProvider.cs:line 32 at System.Composition.Hosting.Providers.Lazy.LazyWithMetadataExportDescriptorProvider.GetExportDescriptors(CompositionContract exportKey, DependencyAccessor definitionAccessor) in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Hosting\src\System\Composition\Hosting\Providers\Lazy\LazyWithMetadataExportDescriptorProvider.cs:line 27 at System.Composition.Hosting.Core.ExportDescriptorRegistryUpdate.GetPromises(CompositionContract contract) in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Hosting\src\System\Composition\Hosting\Core\ExportDescriptorRegistryUpdate.cs:line 148 at System.Composition.Hosting.Core.DependencyAccessor.TryResolveOptionalDependency(Object site, CompositionContract contract, Boolean isPrerequisite, CompositionDependency& dependency) in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Hosting\src\System\Composition\Hosting\Core\DependencyAccessor.cs:line 64 at System.Composition.Hosting.Core.ExportDescriptorRegistryUpdate.Execute(CompositionContract contract) in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Hosting\src\System\Composition\Hosting\Core\ExportDescriptorRegistryUpdate.cs:line 39 at System.Composition.Hosting.Core.ExportDescriptorRegistry.TryGetSingleForExport(CompositionContract exportKey, ExportDescriptor& defaultForExport) in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Hosting\src\System\Composition\Hosting\Core\ExportDescriptorRegistry.cs:line 32 at System.Composition.Hosting.Core.LifetimeContext.TryGetExport(CompositionContract contract, Object& export) in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Hosting\src\System\Composition\Hosting\Core\LifetimeContext.cs:line 197 at System.Composition.Hosting.CompositionHost.TryGetExport(CompositionContract contract, Object& export) in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Hosting\src\System\Composition\Hosting\CompositionHost.cs:line 77 at System.Composition.CompositionContext.GetExport(CompositionContract contract) in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Runtime\src\System\Composition\CompositionContext.cs:line 151 at System.Composition.CompositionContext.GetExport(Type exportType, String contractName) in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Runtime\src\System\Composition\CompositionContext.cs:line 138 at System.Composition.CompositionContext.GetExport[TExport](String contractName) in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Runtime\src\System\Composition\CompositionContext.cs:line 51 at System.Composition.CompositionContext.GetExport[TExport]() in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Runtime\src\System\Composition\CompositionContext.cs:line 38 at System.Composition.UnitTests.MetadataTests.MultiplePiecesOfMetadataAreCombinedIntoAnArray() in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition\tests\MetadataTests.cs:line 127MESSAGE:
System.IO.FileNotFoundException : Could not load file or assembly 'System.Reflection.Emit.Lightweight, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
+++++++++++++++++++
STACK TRACE:
at System.Linq.Expressions.Compiler.LambdaCompiler..ctor(AnalyzedTree tree, LambdaExpression lambda) at System.Linq.Expressions.Compiler.LambdaCompiler.Compile(LambdaExpression lambda) at System.Linq.Expressions.Expression`1.Compile() at System.Composition.Hosting.Providers.Metadata.MetadataViewProvider.GetMetadataViewProvider[TMetadata]() in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Hosting\src\System\Composition\Hosting\Providers\Metadata\MetadataViewProvider.cs:line 73 at System.Composition.Hosting.Providers.Lazy.LazyWithMetadataExportDescriptorProvider.GetLazyDefinitions[TValue,TMetadata](CompositionContract lazyContract, DependencyAccessor definitionAccessor) in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Hosting\src\System\Composition\Hosting\Providers\Lazy\LazyWithMetadataExportDescriptorProvider.cs:line 32 at System.Composition.Hosting.Providers.Lazy.LazyWithMetadataExportDescriptorProvider.GetExportDescriptors(CompositionContract exportKey, DependencyAccessor definitionAccessor) in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Hosting\src\System\Composition\Hosting\Providers\Lazy\LazyWithMetadataExportDescriptorProvider.cs:line 27 at System.Composition.Hosting.Core.ExportDescriptorRegistryUpdate.GetPromises(CompositionContract contract) in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Hosting\src\System\Composition\Hosting\Core\ExportDescriptorRegistryUpdate.cs:line 148 at System.Composition.Hosting.Core.DependencyAccessor.TryResolveOptionalDependency(Object site, CompositionContract contract, Boolean isPrerequisite, CompositionDependency& dependency) in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Hosting\src\System\Composition\Hosting\Core\DependencyAccessor.cs:line 64 at System.Composition.Hosting.Core.ExportDescriptorRegistryUpdate.Execute(CompositionContract contract) in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Hosting\src\System\Composition\Hosting\Core\ExportDescriptorRegistryUpdate.cs:line 39 at System.Composition.Hosting.Core.ExportDescriptorRegistry.TryGetSingleForExport(CompositionContract exportKey, ExportDescriptor& defaultForExport) in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Hosting\src\System\Composition\Hosting\Core\ExportDescriptorRegistry.cs:line 32 at System.Composition.Hosting.Core.LifetimeContext.TryGetExport(CompositionContract contract, Object& export) in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Hosting\src\System\Composition\Hosting\Core\LifetimeContext.cs:line 197 at System.Composition.Hosting.CompositionHost.TryGetExport(CompositionContract contract, Object& export) in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Hosting\src\System\Composition\Hosting\CompositionHost.cs:line 77 at System.Composition.CompositionContext.GetExport(CompositionContract contract) in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Runtime\src\System\Composition\CompositionContext.cs:line 151 at System.Composition.CompositionContext.GetExport(Type exportType, String contractName) in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Runtime\src\System\Composition\CompositionContext.cs:line 138 at System.Composition.CompositionContext.GetExport[TExport](String contractName) in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Runtime\src\System\Composition\CompositionContext.cs:line 51 at System.Composition.CompositionContext.GetExport[TExport]() in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition.Runtime\src\System\Composition\CompositionContext.cs:line 38 at System.Composition.UnitTests.MetadataTests.MultiplePiecesOfMetadataAreCombinedIntoAnArray() in d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Composition\tests\MetadataTests.cs:line 127
```

stephentoub commented on src/System.Composition.Convention/src/System/Composition/Convention/PartConventionBuilder.cs:
```diff
@@ -0,0 +1,822 @@
+    /// </summary>
+    public class PartConventionBuilder
+    {
+        private readonly Type[] _emptyTypeArray = new Type[0];
```
How about using ```Array.Empty<Type>``` at call sites instead of having this field?

dsplaisted replied:
@stephentoub This library targets platforms like .NET 4.5, Windows Store 8, and Windows Phone 8 which I don't think have support for `Array.Empty<T>`.

stephentoub replied:
I see, I didn't realize these libraries weren't just .NET Core versions.

stephentoub commented on src/System.Composition.Hosting/src/System/Composition/Hosting/CompositionHost.cs:
```diff
@@ -0,0 +1,88 @@
+    /// </summary>
+    public sealed class CompositionHost : CompositionContext, IDisposable
+    {
+        private static readonly string[] s_noBoundaries = new string[0];
```
How about using ```Array.Empty<string>``` at call sites instead of having this field?

stephentoub commented on src/System.Composition.Hosting/src/System/Composition/Hosting/Core/ExportDescriptorRegistryUpdate.cs:
```diff
@@ -0,0 +1,156 @@
+        private readonly ExportDescriptorProvider[] _exportDescriptorProviders;
+        private readonly IDictionary<CompositionContract, UpdateResult> _updateResults = new Dictionary<CompositionContract, UpdateResult>();
+
+        private static readonly CompositionDependency[] s_noDependenciesValue = new CompositionDependency[0];
```
```Array.Empty<CompositionDependency>```?

stephentoub commented on src/System.Composition.Hosting/src/System/Composition/Hosting/Core/UpdateResult.cs:
```diff
@@ -0,0 +1,59 @@
+    // providers, with providers being removed from the list before querying.
+    internal class UpdateResult
+    {
+        private static readonly ExportDescriptorPromise[] s_noPromises = new ExportDescriptorPromise[0];
```
```Array.Empty<ExportDescriptorPromise>``` ?

stephentoub commented on src/System.Composition.Hosting/src/System/Composition/Hosting/Providers/ExportFactory/ExportFactoryExportDescriptorProvider.cs:
```diff
@@ -0,0 +1,64 @@
+        private static ExportDescriptorPromise[] GetExportFactoryDescriptors<TProduct>(CompositionContract exportFactoryContract, DependencyAccessor definitionAccessor)
+        {
+            var productContract = exportFactoryContract.ChangeType(typeof(TProduct));
+            var boundaries = new string[0];
```
```Array.Empty<string>``` ?

stephentoub replied:
Even if there's no ```Array.Empty<T>``` available, you can still fix cases like this, either by having it use a static readonly field that caches an instance (e.g. CompositionHost.s_noBoundaries), or by having your own ```EmptyArray<T>.Instance``` that you define and then use in all of these places.

dsplaisted replied:
@stephentoub I've added `EmptyArray<T>` to the Common folder, and I think I've updated all instances of creating an empty array to use it.

stephentoub commented on src/System.Composition.Hosting/src/System/Composition/Hosting/Providers/ExportFactory/ExportFactoryExportDescriptorProvider.cs:
```diff
@@ -0,0 +1,64 @@
+            if (exportFactoryContract.TryUnwrapMetadataConstraint(Constants.SharingBoundaryImportMetadataConstraintName, out specifiedBoundaries, out unwrapped))
+            {
+                productContract = unwrapped.ChangeType(typeof(TProduct));
+                boundaries = (specifiedBoundaries ?? new string[0]).ToArray();
```
```Array.Empty<string>``` ?

stephentoub commented on src/System.Composition.Hosting/src/System/Composition/Hosting/Providers/ExportFactory/ExportFactoryWithMetadataExportDescriptorProvider.cs:
```diff
@@ -0,0 +1,71 @@
+        private static ExportDescriptorPromise[] GetExportFactoryDescriptors<TProduct, TMetadata>(CompositionContract exportFactoryContract, DependencyAccessor definitionAccessor)
+        {
+            var productContract = exportFactoryContract.ChangeType(typeof(TProduct));
+            var boundaries = new string[0];
```
```Array.Empty<string>``` ?

stephentoub commented on src/System.Composition.Hosting/src/System/Composition/Hosting/Providers/ExportFactory/ExportFactoryWithMetadataExportDescriptorProvider.cs:
```diff
@@ -0,0 +1,71 @@
+            if (exportFactoryContract.TryUnwrapMetadataConstraint(Constants.SharingBoundaryImportMetadataConstraintName, out specifiedBoundaries, out unwrapped))
+            {
+                productContract = unwrapped.ChangeType(typeof(TProduct));
+                boundaries = (specifiedBoundaries ?? new string[0]).ToArray();
```
```Array.Empty<string>``` ?

stephentoub commented on src/System.Composition.TypedParts/src/System/Composition/TypedParts/ActivationFeatures/ActivationFeature.cs:
```diff
@@ -0,0 +1,46 @@
+        /// <summary>
+        /// A constant shared for subclass use.
+        /// </summary>
+        protected static readonly CompositionDependency[] NoDependencies = new CompositionDependency[0];
```
```Array.Empty<CompositionDependency>``` ?

stephentoub commented on src/System.Composition.TypedParts/src/System/Composition/TypedParts/Util/DirectAttributeContext.cs:
```diff
@@ -0,0 +1,31 @@
+            if (member == null) throw new ArgumentNullException("member");
+
+            if (!(member is TypeInfo) && member.DeclaringType != reflectedType)
+                return new Attribute[0];
```
```Array.Empty<Attribute>``` ?

KrzysztofCwalina commented on src/System.Composition.Convention/src/Microsoft/Internal/Requires.cs:
```diff
@@ -0,0 +1,104 @@
+namespace Microsoft.Internal
+{
+    internal static class Requires
+    {
```
Above I say a type called "Assume" that was very similar. Do we need to have both?

dsplaisted replied:
@KrzysztofCwalina I think `Requires` is meant to specify preconditions (which someone calling your code may violate), and 'Assumes" is meant to specify invariants, which if they fail means there's a bug in the library code (in this case System.Composition).

stephentoub replied:
If that's the case, should we just use Debug.Assert instead of Assumes?  That's how invariants are handles in most of the other libraries in corefx.

KrzysztofCwalina commented on src/System.Composition.Convention/src/System/Composition/Convention/ConventionBuilder.cs:
```diff
@@ -0,0 +1,351 @@
+    /// Entry point for defining rules that configure plain-old-CLR-objects as MEF parts.
+    /// </summary>
+    public class ConventionBuilder : AttributedModelProvider
+    {
```
Would it make sense for all these convention based builders to be in a separate assembly?

dsplaisted replied:
@KrzysztofCwalina I don't really follow.  They are already the only public types exposed by System.Composition.Convention, as far as I can see.

KrzysztofCwalina commented on src/System.Composition.Convention/src/System/Composition/ExceptionBuilder.cs:
```diff
@@ -0,0 +1,42 @@
+
+namespace System.Composition
+{
+    internal static class ExceptionBuilder
```
What value does this type add?

mellinoe commented on src/System.Composition.AttributedModel/src/System.Composition.AttributedModel.csproj:
```diff
@@ -0,0 +1,49 @@
+    <RootNamespace>System.Composition</RootNamespace>
+    <AssemblyName>System.Composition.AttributedModel</AssemblyName>
+    <TargetFrameworkVersion>v4.5</TargetFrameworkVersion>
+    <TargetFrameworkProfile>Profile259</TargetFrameworkProfile>
```
Is this how we intend to continue building this library? As far as I'm aware this would be the only library that actually builds against a portable profile rather than with our contract assemblies via NuGet. We don't have a way to compile against the set of contracts exposed by this profile (lower versions), so moving to NuGet would prevent us from running on Win8/WP8. Just throwing this out there.

dsplaisted replied:
@mellinoe System.Collections.Immutable supports the same set of platforms as System.Composition.

akoeplinger commented on src/System.Composition.Convention/tests/System.Composition.Convention.Tests.csproj:
```diff
@@ -0,0 +1,85 @@
+    </ProjectReference>
+  </ItemGroup>
+  <ItemGroup>
+    <None Include="packages.config">
```
This doesn't exist anymore.

akoeplinger commented on src/System.Composition/tests/System.Composition.Tests.csproj:
```diff
@@ -0,0 +1,105 @@
+    </ProjectReference>
+  </ItemGroup>
+  <ItemGroup>
+    <None Include="packages.config.bak">
```
This doesn't exist anymore.

akoeplinger commented on src/System.Composition/perftests/project.json:
```diff
@@ -0,0 +1,18 @@
+    "System.Runtime.Extensions": "4.0.10-beta-22809",
+    "System.Threading": "4.0.10-beta-22809",
+    "System.Threading.Tasks.Parallel": "4.0.0-beta-22809",
+    "System.Threading.Thread": "4.0.0-beta-22809"
```
Should use `-*` versions like the rest of the projects.

akoeplinger commented on src/System.Composition.sln:
```diff
@@ -0,0 +1,97 @@
+﻿
```
I'm not a fan of keeping this directly in \src, all the other solution files are inside a subfolder and it looks much cleaner. Can we move this to \src\System.Composition?

dsplaisted replied:
Would you just move the .sln file, or would you move all the System.Composition code under a common \src\System.Composition folder?

akoeplinger replied:
All the other folders in \src seem to follow the "folder == nuget package" pattern, so if the assemblies ship in separate packages then it's fine as is with multiple folders (with the sln being in \src\System.Composition as I think that's the most logical place).

dsplaisted replied:
For all the other projects, each NuGet package corresponds to a single assembly (AFAIK).  For System.Composition, right now all the assemblies are shipped in [one NuGet package](http://www.nuget.org/packages/Microsoft.Composition/).  So it doesn't really fit the existing pattern.

akoeplinger replied:
Hmm, if the assemblies are shipped together then I'd move everything into a common src\System.Composition as to me the "folder == nuget package" convention is more important than "nuget package == single assembly".

akoeplinger commented on src/System.Composition.TypedParts/src/System.Composition.TypedParts.csproj:
```diff
@@ -0,0 +1,84 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="4.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
We standardized on `ToolsVersion="12.0"` in the other projects. This applies to all the .csproj files in this PR.

akoeplinger commented on src/System.Composition.sln:
```diff
@@ -0,0 +1,97 @@
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "System.Composition.Convention.Tests", "System.Composition.Convention\tests\System.Composition.Convention.Tests.csproj", "{853BB14F-8A5B-42B4-A053-21DE1AEBB335}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Microsoft.Composition.Demos.ExtendedCollectionImports", "System.Composition\demos\Microsoft.Composition.Demos.ExtendedCollectionImports\Microsoft.Composition.Demos.ExtendedCollectionImports.csproj", "{44C7E52C-3873-4C64-875C-8A23A8376D60}"
```
I'm wondering if this is a real "demo" project (as in something that a user would look at to learn about System.Composition)? I just see it is used in some unit tests, maybe a different name would be better?

mmitche commented:
@dotnet-bot test this please

stephentoub commented:
@dsplaisted, what's the status of this?

dsplaisted commented:
@stephentoub I haven't had time to work on this recently.  I need to investigate why the CI build is failing and go back and see if I've responded to all the feedback.

dsplaisted commented:
@dotnet-bot test this please

jasonwilliams200OK commented:
It seems like only Windows Debug build is failing on running `System.Composition` tests due to missing dependency: `System.Reflection.Emit.Lightweight`

> System.IO.FileNotFoundException : Could not load file or assembly 'System.Reflection.Emit.Lightweight, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified

stephentoub commented:
@dsplaisted, it's been a month, so I wanted to check in on this again.  Have you been able to make any progress?  Do you need assistance?

dsplaisted commented:
I think I've fixed all the build and test failures.

I've added checklists in the original PR description for outstanding issues, as well as a list of things to file bugs for once the PR is merged.  Other than that, I think I've addressed all the feedback.  Let me know if I missed anything!

dsplaisted commented:
In regard to [unnecessary using statements](https://github.com/dotnet/corefx/pull/1851#discussion_r31222278), @stephentoub wrote:

> I view unused using as dead code, and dead code should be removed. Doesn't seem like a style thing to me, personally.

The templates when you create a new code file add some default using statements for you for namespaces you are likely to use.  I would normally leave those using statements there- if you go back and modify the code you may use types from those namespaces.  (Yes, we have nice tooling to add using statements for you, but you still don't get intellisense for APIs until you've typed the full name (correctly) and used the quick fix to add the using statement.)  And now that with VS 2015 we can use LINQ queries in the watch and immediate windows, a using statement for System.Linq is especially useful, as it will let you use LINQ extension methods when debugging even if your original code didn't use them.

So if we don't want any unnecessary using statements, I think we should put it in the style guide.

mmitche commented:
@dsplaisted Ignore that last failure.  I'm adding in the native component builds to the corefx testing and your change picked up the copy side of things without the corresponding build.  I'm restarting it.

mmitche commented:
@dotnet-bot test this please

stephentoub commented:
> The templates when you create a new code file add some default using statements for you for namespaces you are likely to use

That's true.  At the same time, in VS2015 unused namespaces are dimmed out and Quick Actions push you towards eliminating all used namespaces.

stephentoub commented:
Thanks, Daniel!  Glad it's all building and almost ready to go. I'll re-review one more time once the remaining issues are addressed and it's squashed down (doesn't need to be a single commit, but it should be many fewer than 33, especially since many of those commits didn't build successfully).

jasonwilliams200OK commented:
:+1:

Maxwe11 commented on src/System.Composition.Convention/src/System.Composition.Convention.csproj:
```diff
@@ -0,0 +1,88 @@
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
+  </PropertyGroup>
+  <ItemGroup>
+    <Compile Include="..\..\Common\src\Microsoft\Internal\Assumes.cs">
```
`"..\..\Common\src"` should be replaced with `"$(CommonPath)"`

stephentoub commented:
Regarding the remaining issues:

> Are AssemblyInfo.cs files needed?

No.

> Remove unused usings?

Ideally, but there are tons of unused usings across corefx as is.  I'm ok with this being addressed separately and in a cleanup effort consistently across the repo.

> Dependencies in project.json should be consistent with rest of repo (ie use "-beta-*")

At this point, stable dependencies no longer need "-beta-*".

stephentoub commented on src/System.Composition.Convention/tests/project.lock.json:
```diff
@@ -0,0 +1,1185 @@
+{
+  "locked": false,
```
Needs to be true

stephentoub commented on src/System.Composition/perftests/project.lock.json:
```diff
@@ -0,0 +1,630 @@
+{
+  "locked": false,
```
Needs to be true

stephentoub commented on src/System.Composition/tests/project.lock.json:
```diff
@@ -0,0 +1,1142 @@
+{
+  "locked": false,
```
Needs to be true

stephentoub commented on src/Common/src/Microsoft/Internal/Assumes.cs:
```diff
@@ -0,0 +1,114 @@
+        }
+
+        [DebuggerStepThrough]
+        internal static void IsTrue(bool condition, /*[Localizable(false)]*/string message)
```
What's the plan for these commented-out ```[Localizable(...)]``` attributes?

dsplaisted replied:
> What's the plan for these commented-out ```[Localizable(...)]``` attributes?

I'm not sure- perhaps at the time this was being written the attribute wasn't portable.  Can you shine any light on this, @nblumhardt?

nblumhardt replied:
@dsplaisted I would guess it was something like that. `Assumes` came across wholesale from some of the "classic" MEF tests. Can't see any reason not to uncomment them.

stephentoub commented on src/System.Composition.AttributedModel/src/System.Composition.AttributedModel.csproj:
```diff
@@ -0,0 +1,49 @@
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
+  </PropertyGroup>
+  <ItemGroup>
+    <!-- A reference to the entire .NET Framework is automatically included -->
```
This ItemGroup can be deleted; the comment no longer looks correct.

stephentoub commented on src/System.Composition.Hosting/src/System/Composition/Hosting/Providers/Constants.cs:
```diff
@@ -0,0 +1,27 @@
+        /// <summary>
+        /// The sharing boundary implemented by an import.
+        /// </summary>
+        public const string SharingBoundaryImportMetadataConstraintName = "SharingBoundaryNames";
```
This string also appeared in one of the other assemblies.  Should this file be in some shared location and brought in to each project that needs it, so that the string need only be written out in one place?

dsplaisted replied:
The other copy of it was a private constant that was never used, so I've deleted it.

stephentoub commented:
Thanks, Daniel.  Left a few more comments, but otherwise LGTM for the initial merge.

dsplaisted commented:
I've addressed all the feedback.  I think @joshfree was of the opinion that when adding newly ported code, it should all be squashed down to one commit.  Should I do so and then press the big green "merge" button?

stephentoub commented:
> I think @joshfree was of the opinion that when adding newly ported code, it should all be squashed down to one commit. Should I do so and then press the big green "merge" button?

Personally I'd suggest squashing this down to two commits: one that covers your first three commits currently (initial code, formatting, building), and one that covers all of the subsequent cleanup you then did.  The first commit would be from dotnet-bot, and the second commit would be from you. But if you feel there's a cleaner approach, maybe you feel some specific commit should be kept separate, etc., that's fine, too.

Once that's done and CI passes again, merge away.

jasonwilliams200OK commented:
So now it is just a matter of squashing the commits and after that it will be merged? Yay! Love to see System.Composition getting merged into CoreFX! :fireworks: 

Going by @stephentoub's suggestion and other "initial commit" PRs, it would probably be 1c3415a, 1f1d240, 63f15b5 and 62bc5bd fixed up to 5b5364c. And remaining squashed to a separate commit with bulleted list (`* my message`) of changes in the commit log.

Thanks for doing this @dsplaisted! :ok_hand:
